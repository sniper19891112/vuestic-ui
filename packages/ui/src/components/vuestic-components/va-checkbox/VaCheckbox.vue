<template>
  <va-input-wrapper
    class="va-checkbox"
    :class="computedClass"
    :disabled="c_disabled"
    :success="c_success"
    :messages="c_messages"
    :error="computedError"
    :error-messages="computedErrorMessages"
    :error-count="errorCount"
  >
    <div
      class="va-checkbox__input-container"
      @click="onWrapperClick()"
      @mousedown="hasMouseDown = true"
      @mouseup="hasMouseDown = false"
      tabindex="-1"
      @blur="onBlur"
      ref="container"
    >
      <div
        class="va-checkbox__square"
        :style="inputStyle"
      >
        <input
          ref="input"
          :id="id"
          :name="name"
          type="checkbox"
          role="checkbox"
          readonly
          @focus="onFocus"
          @blur="onBlur($event)"
          class="va-checkbox__input"
          @keypress.prevent="toggleSelection()"
          :disabled="c_disabled"
          :indeterminate="c_indeterminate"
        >
        <va-icon
          class="va-checkbox__icon"
          :name="computedIconName"
          size="20px"
        />
      </div>
      <div
        class="va-checkbox__label"
        :style="labelStyle"
        ref="label"
        tabindex="-1"
        @blur="onBlur"
      >
        <slot name="label">
          {{ c_label }}
        </slot>
      </div>
    </div>
  </va-input-wrapper>
</template>

<script>
import VaIcon from '../va-icon/VaIcon'
import { SelectableMixin } from '../../vuestic-mixins/SelectableMixin/SelectableMixin'
import { makeContextablePropsMixin } from '../../context-test/context-provide/ContextPlugin'
import { getColor } from '../../../services/ColorThemePlugin'
import VaInputWrapper from '../va-input/VaInputWrapper'

export default {
  name: 'VaCheckbox',
  components: { VaInputWrapper, VaIcon },
  mixins: [
    SelectableMixin,
    makeContextablePropsMixin({
      value: { type: [Boolean, Array, String, Object], default: false },
      checkedIcon: { type: String, default: 'check' },
      indeterminateIcon: { type: String, default: 'remove' },
      color: { type: String, default: 'primary' },
    }),
  ],
  computed: {
    computedClass () {
      return {
        'va-checkbox--selected': this.isChecked,
        'va-checkbox--readonly': this.c_readonly,
        'va-checkbox--disabled': this.c_disabled,
        'va-checkbox--indeterminate': this.c_indeterminate,
        'va-checkbox--error': this.computedError,
        'va-checkbox--left-label': this.c_leftLabel,
        'va-checkbox--on-keyboard-focus': this.isKeyboardFocused,
      }
    },
    labelStyle () {
      return {
        color: this.computedError ? getColor(this, 'danger') : '',
        padding: !this.c_label
          ? ''
          : this.c_leftLabel
            ? '0 0.25rem 0 0'
            : '0 0 0 0.25rem',
      }
    },
    inputStyle () {
      return this.computedError
        ? (this.isChecked || this.isIndeterminate)
          ? { background: this.colorComputed, borderColor: getColor(this, 'danger') }
          : { borderColor: getColor(this, 'danger') }
        : (this.isChecked || this.isIndeterminate)
          ? { background: this.colorComputed, borderColor: this.colorComputed }
          : {}
    },
    computedIconName () {
      return (this.c_indeterminate && this.isIndeterminate)
        ? this.c_indeterminateIcon
        : this.c_checkedIcon
    },
  },
}
</script>

<style lang="scss" scoped>
@import "../../vuestic-sass/resources/resources";

.va-checkbox {
  display: flex;
  flex-direction: column;
  justify-content: center;

  &__input-container {
    align-items: center;
    display: flex;
    padding: 0 0.3rem;
    cursor: pointer;

    @at-root {
      .va-checkbox--disabled & {
        @include va-disabled();

        cursor: default;
      }

      .va-checkbox--readonly & {
        cursor: initial;
      }

      .va-checkbox--left-label & {
        flex-direction: row-reverse;
      }
    }
  }

  #{&}__square {
    @include flex-center();

    width: 1.35rem;
    min-width: 1.35rem;
    height: 1.35rem;
    position: relative;
    background-color: $white;
    border: solid 0.125rem $gray-light;
    border-radius: 0.25rem;

    @at-root {
      .va-checkbox--on-keyboard-focus#{&} {
        transition: all, 0.6s, ease-in;
        box-shadow: 0 0 0.5rem 0 rgba(0, 0, 0, 0.3);
      }
    }
  }

  &__input {
    opacity: 0;
    width: 0;
    height: 0;
  }

  &__label {
    display: inline-block;
    position: relative;
  }

  &__icon {
    pointer-events: none;
    position: absolute;
    color: transparent;
  }

  &--selected {
    .va-checkbox {
      &__icon {
        color: $white;
      }
    }
  }

  &--indeterminate {
    .va-checkbox {
      &__icon {
        color: $white;
      }
    }
  }
}

</style>
