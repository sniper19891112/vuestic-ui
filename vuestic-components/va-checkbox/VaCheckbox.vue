<template>
  <div
    class="va-checkbox"
    :class="computedClass"
  >
    <div
      class="va-checkbox__input-container"
      @click="toggleSelection()"
      @mousedown="onMouseDown"
      @mouseup="onMouseUp"
    >
      <div
        class="va-checkbox__square"
        :class="{'active': value}"
      >
        <input
          :id="id"
          readonly
          @focus="onFocus"
          @blur="isKeyboardFocused = false"
          class="va-checkbox__input"
          @keypress.prevent="toggleSelection()"
          :disabled="disabled"
        />
        <va-icon icon="ion ion-md-checkmark va-checkbox__icon-selected"/>
      </div>
      <div class="va-checkbox__label-text">
        <slot name="label">
          {{ label }}
        </slot>
      </div>
    </div>
    <div class="va-checkbox__error-message-container" v-if="showError">
      <div
        class="va-checkbox__error-message"
        v-for="(error, index) in computedErrorMessages"
        :key="index"
      >
        {{ error }}
      </div>
    </div>
  </div>
</template>

<script>
import VaIcon from '../va-icon/VaIcon'

export default {
  name: 'va-checkbox',
  components: { VaIcon },
  props: {
    label: String,
    value: {
      type: Boolean,
      required: true,
    },
    id: {
      type: String,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    errorMessages: {
      type: [String, Array],
      default: () => [],
    },
    errorCount: {
      type: Number,
      default: 1,
    },
    name: String,
    error: {
      type: Boolean,
      default: false,
    },
  },
  data () {
    return {
      isKeyboardFocused: false,
      hasMouseDown: false,
    }
  },
  computed: {
    computedClass () {
      return {
        'va-checkbox--selected': this.value,
        'va-checkbox--readonly': this.readonly,
        'va-checkbox--disabled': this.disabled,
        'va-checkbox--error': this.showError,
        'va-checkbox--on-keyboard-focus': this.isKeyboardFocused,
      }
    },
    computedErrorMessages () {
      const isArray = Array.isArray(this.errorMessages)
      const errorMessages = isArray ? this.errorMessages : [this.errorMessages]
      return errorMessages.slice(0, this.errorCount)
    },
    valueProxy: {
      set (value) {
        this.$emit('input', value)
      },
      get () {
        return this.value
      },
    },
    showError () {
      // We make error active, if the error-message is not empty and checkbox is not disabled
      if (!this.disabled) {
        if (!(this.errorMessages.length === 0) || this.error) {
          return true
        }
      }
      return false
    },
  },
  methods: {
    onFocus (e) {
      if (this.hasMouseDown) {
        return
      }
      this.isKeyboardFocused = true
    },
    onMouseDown (e) {
      this.hasMouseDown = true
      this.$emit('mousedown', e)
    },
    onMouseUp (e) {
      this.hasMouseDown = false
      this.$emit('mouseup', e)
    },
    toggleSelection () {
      if (this.readonly) {
        return
      }
      if (this.disabled) {
        return
      }
      this.valueProxy = !this.valueProxy
    },
  },
}
</script>

<style lang="scss">
@import "../../vuestic-sass/resources/resources";

.va-checkbox {
  margin-bottom: $checkbox-between-items-margin;
  display: flex;
  flex-direction: column;

  &__input-container {
    align-items: center;
    display: flex;
    cursor: pointer;

    @at-root {
      .va-checkbox--disabled  & {
        @include va-disabled();
      }

      .va-checkbox--readonly & {
        cursor: initial;
      }

      .va-checkbox--disabled & {
        cursor: default;
      }
    }
  }

  #{&}__square {
    @include flex-center();
    width: 2rem;
    height: 2rem;
    position: relative;
    flex: 0 0 2rem;
    @at-root {
      .va-checkbox--on-keyboard-focus#{&} {
        background-color: $light-gray;
        transition: all, 0.6s, ease-in;
        border-radius: 5rem;
      }
    }
  }

  #{&}__input {
    height: 1.375rem;
    width: 1.375rem;
    cursor: inherit;
    color: $white;
    background-color: $white;
    border: solid 0.125rem $gray-light;
    border-radius: 0.25rem;

    &:focus {
      outline: none;
    }

    @at-root {
      .va-checkbox--selected#{&} {
        display: flex;
        justify-content: center;
        align-items: center;
        height: 1.4rem;
        width: 1.4rem;
        color: $white;
        background-color: $vue-green;
        border: 0;
      }

      .va-checkbox--error#{&} {
        border-color: $theme-red;
      }
    }
  }

  #{&}__label-text {
    display: inline-block;
    position: relative;
    margin-left: 0.25rem;
    @at-root {
      .va-checkbox--error#{&} {
        color: $theme-red;
      }
    }
  }

  &__error-message {
    vertical-align: middle;
    color: $theme-red;
    font-size: $font-size-mini;
  }

  &__icon-selected {
    pointer-events: none;
    position: absolute;
    color: $white;
  }

  &__error-message-container {
    flex: 0 0 100%;
    margin-left: 0.3rem; // To fit with checkbox.
  }

  &__label-container {
    margin-left: 2rem;
  }

  &__content {
    flex-direction: row;
  }
}
</style>
