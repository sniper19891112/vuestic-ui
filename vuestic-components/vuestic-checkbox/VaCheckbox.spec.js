import { shallowMount } from '@vue/test-utils'
import VaCheckbox from './VuesticCheckbox'

describe('VaCheckbox', () => {
  it('default', () => {
    const wrapper = shallowMount(VaCheckbox, {
      propsData: { value: false }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('true value', () => {
    const wrapper = shallowMount(VaCheckbox, {
      propsData: { value: true }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
