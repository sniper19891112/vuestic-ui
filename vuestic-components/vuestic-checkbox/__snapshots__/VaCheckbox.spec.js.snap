// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaCheckbox default 1`] = `
<div class="vuestic-checkbox">
  <div class="vuestic-checkbox__square"><input readonly="readonly" class="vuestic-checkbox__input">
    <va-icon-stub icon="ion ion-md-checkmark vuestic-checkbox__icon-selected"></va-icon-stub>
  </div>
  <div class="vuestic-checkbox__label-text">

  </div>
  <!---->
</div>
`;

exports[`VaCheckbox true value 1`] = `
<div class="vuestic-checkbox vuestic-checkbox--selected">
  <div class="vuestic-checkbox__square active"><input readonly="readonly" class="vuestic-checkbox__input">
    <va-icon-stub icon="ion ion-md-checkmark vuestic-checkbox__icon-selected"></va-icon-stub>
  </div>
  <div class="vuestic-checkbox__label-text">

  </div>
  <!---->
</div>
`;
